### PR TITLE
feat: validate barcalendar project versions dates from eol

### PR DESCRIPTION
**Pull request by UsamaSadiq**
## Description
- PR created under the issue https://github.com/edx/edx-arch-experiments/issues/199

## Changes 
- Added new validation function `validate_version_date` for validating a `project/version`'s EOL or release Date from https://endoflife.date/docs/api 
- Some Projects like `Django` & `Ubuntu` don't have any EOL in our script and we only validate their release dates so added the option to validate either release date or EOL based on a boolean flag input. 
- Added custom logger to handle all console logs separately from the console script being generated. 

## Current log Output
- Currently following issues are being reported when running the script.

```
<LogRecord: eol_logger, 30, barcalendar.py, 100, "Invalid EOL: Update Django 3.0 releaseDate to 2019/12 instead">
<LogRecord: eol_logger, 30, barcalendar.py, 100, "Invalid EOL: Update Django 4.0 releaseDate to 2021/12 instead">
<LogRecord: eol_logger, 40, barcalendar.py, 83, "Project not Found: Ubuntu/24.04 not found on endoflife.date">
<LogRecord: eol_logger, 40, barcalendar.py, 83, "Project not Found: Ubuntu/26.04 not found on endoflife.date">
<LogRecord: eol_logger, 30, barcalendar.py, 100, "Invalid EOL: Update MySQL 8.1 eol to 2023/10 instead">
<LogRecord: eol_logger, 40, barcalendar.py, 83, "Project not Found: ElasticSearch/7.10 not found on endoflife.date">
<LogRecord: eol_logger, 40, barcalendar.py, 83, "Project not Found: ElasticSearch/7.11 not found on endoflife.date">
<LogRecord: eol_logger, 40, barcalendar.py, 83, "Project not Found: ElasticSearch/7.12 not found on endoflife.date">
<LogRecord: eol_logger, 40, barcalendar.py, 83, "Project not Found: ElasticSearch/7.13 not found on endoflife.date">
<LogRecord: eol_logger, 40, barcalendar.py, 83, "Project not Found: ElasticSearch/7.17 not found on endoflife.date">
<LogRecord: eol_logger, 40, barcalendar.py, 83, "Project not Found: Redis/5.6 not found on endoflife.date">
<LogRecord: eol_logger, 30, barcalendar.py, 100, "Invalid EOL: Update Redis 6.0 eol to 2023/08 instead">
<LogRecord: eol_logger, 30, barcalendar.py, 92, "No EOL found for Redis 6.2">
<LogRecord: eol_logger, 30, barcalendar.py, 92, "No EOL found for Redis 7.0">

```

## Automation of the Warning Resolver Script
- There is an automated [workflow](https://github.com/openedx/repo-tools/blob/master/.github/workflows/support-window-issue-creator.yml) already working which executes on quarterly basis and creates an issue for the arbi-bom team to update support window sheet. (Most recent [issue](https://github.com/openedx/repo-tools/issues/445) created on Oct 1st for current quarter). Above warnings will be fixed in resolving this issue.

### CODE REVIEW

File: barcalendar.py [hunk 1]
```diff
@@ -22,9 +22,87 @@
 import re
 import time
 
+import atexit
+import logging
```
Comment (nedbat) [inline barcalendar.py:26]: [nit] These two lines from the stdlib, so should be alphabetized into the first block of imports.

File: barcalendar.py [hunk 2]
```diff
@@ -22,9 +22,87 @@
...
+            if version_info['eol'] == False:
+                eol_logger.warning(f"No EOL found for {project} {version}")
+                return (year, month)                
+            eol_year, eol_month, _ = version_info['eol'].split("-")
+            date_type = "eol"
+        
+        if (year, month) == (int(eol_year), int(eol_month)):
+            return (year, month)
+        else:
+            eol_logger.warning(f"Invalid EOL: Update {project} {version} {date_type} to {eol_year}/{eol_month} instead")
```
Comment (nedbat) [inline barcalendar.py:100]: [nit] year/month is an unusual format.  Use month/year, or year-month.

File: barcalendar.py [hunk 3]
```diff
@@ -22,9 +22,87 @@
...
 import time
 
+import atexit
+import logging
 import requests
 import yaml
 
+def setup_custom_logging():
+    """
+    Creating a custom logger to flush all Warning logs at the end of the script
```
Comment (nedbat) [inline barcalendar.py:32]:
```suggestion
    Create a custom logger to flush all Warning logs at the end of the script
```

File: barcalendar.py [hunk 4]
```diff
@@ -22,9 +22,87 @@
...
+        def __init__(self):
+            super().__init__()
+            self.log_messages = []
+
+        def emit(self, record):
+            self.log_messages.append(record)
+
+    def _flush_log_messages(handler):
+        for message in handler.log_messages:
+            print(message)
```
Comment (nedbat) [inline barcalendar.py:45]: If you prefix the message with `// ` then the full output will still be valid JavaScript, and could be copied into the spreadsheet without worrying about the log messages.

Also, these are records, not messages.  IDK if you want to extract the message to have a more readable output.

**Review by nedbat — CHANGES_REQUESTED** [review]

**Comment by nedbat** [comment]
Go ahead and merge after small tweaks.

**Review by nedbat — APPROVED** [review]

File: barcalendar.py (hunk 1 above)
Comment (UsamaSadiq) [reply barcalendar.py:26]: Done.

File: barcalendar.py (hunk 2 above)
Comment (UsamaSadiq) [reply barcalendar.py:100]: Done.

File: barcalendar.py (hunk 4 above)
Comment (UsamaSadiq) [reply barcalendar.py:45]: Updated as per the suggestions.

**Pull request merged**